src/oauthhandler.ts

```diff
@@ -0,0 +1,126 @@
+            path: "/_bridge/oauth",
+        });
+        log.info("Bound _bridge/oauth for oAuth");
+    }
+
+    public handleOAuthRequest(userid: string) {
+        // 1. Get the userId
+        const stateKey = uuid();
+        this.stateUserMap.set(stateKey, userid);
+        const url = `${this.authUrl}&state=${stateKey}`;
+        return url;
+    }
+
+    public async onGetOAuth(req: Request, res: Response) {
+        const userId = this.stateUserMap.get(req.query.state);
+        if (!userId) {
+            log.warn("Ignorning oauth request for unknown user");
+            res.status(HTTP_UNAUTHORISED).send("Error: User not found");
+            return;
+        }
+        this.stateUserMap.delete(req.query.state);
+        log.info("Got OAuth Request for", userId);
+        const formData = {
+            client_id: this.config.clientID,
+            client_secret: this.config.clientSecret,
+            code: req.query.code,
+            grant_type: "authorization_code",
+            redirect_uri: this.config.oAuthUrl,
+            scope: this.scopes(),
+        };
+        try {
+            // Tslint fails to recognise these as promises.
+            /* tslint:disable-next-line await-promise */
+            const atRes = JSON.parse(await (request.post({
+                formData,
+                simple: true,
+                url: URL_TOKEN,
+            }))) as IDiscordTokenResponse;
+            const dataObj = new DbAccessToken();
+            dataObj.DiscordId = await this.getIdentity(atRes.access_token);
+            dataObj.AccessToken = atRes.access_token;
+            dataObj.RefreshToken = atRes.refresh_token;
+            dataObj.ExpiresIn = atRes.expires_in || 1;
+            dataObj.MatrixId = userId;
+            await this.store.Insert(dataObj);
+        } catch (ex) {
+            log.warn("Got error when trying to get token:", ex);
+            res.status(HTTP_INTERNAL_ERROR).send("Error: Unknown error");
+            return;
+        }
+        res.send("Your Discord account has been linked to this bridge instance. You may close this page.");
+        // TODO: We could trigger a power level update for any guilds this user is in.
+    }
+
+    private async getIdentity(accessToken: string) {
+        /* tslint:disable-next-line await-promise */
+        const ident = await request.get({
+            headers: {
+                Authorization: `Bearer ${accessToken}`,
+            },
+            json: true,
+            simple: true,
+            url: URL_IDENTITY,
+        });
+        log.verbose("Identified as", ident.id, ident.username);
+        return ident.id;
+    }
+
+    private scopes(escape: boolean = false) {
+        return ["identify", "guilds", "guilds.join"].join(escape ? "%20" : " ");
+    }
+
+    private get authUrl() {
+        return `${URL_AUTH}?client_id=${this.config.clientID}&response_type=code` +
+        `&redirect_uri=${this.config.oAuthUrl}&scope=${this.scopes(true)}`;
```

Comment:
scope doesn't need to be that integer bitfield? It can be easily calculated using discord.js